File: 0109-convert-sorted-list-to-binary-search-tree/0109-convert-sorted-list-to-binary-search-tree.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public TreeNode sortedListToBST(ListNode head) {
        return f(head, null);
    }
    private TreeNode f(ListNode start, ListNode end) {
        if(start == end) return null;
        
        ListNode mid = start, fast = start;
        while(fast != end && fast.next != end) {
            mid = mid.next;
            fast = fast.next.next;
        }
        
        TreeNode currNode = new TreeNode(mid.val);
        currNode.left = f(start, mid);
        currNode.right = f(mid.next, end);
        
        return currNode;
    }
}

File: 0109-convert-sorted-list-to-binary-search-tree/0109-convert-sorted-list-to-binary-search-tree.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* f(ListNode* low, ListNode* high) {
        if(low == high) return nullptr;
        
        // if(low == high) {
        //     TreeNode* new_node = new TreeNode(low->val);
        //     return new_node;
        // }
        
        ListNode* slow = low, *fast = low;
        ListNode* prev = nullptr;
        while(fast != high and fast->next != high) {
            slow = slow->next;
            fast = fast->next->next;
            prev = slow;
        }
        
        TreeNode* new_node = new TreeNode(slow->val);
        new_node->left = f(low, slow);
        new_node->right = f(slow->next, high);
        return new_node;
    }
    TreeNode* sortedListToBST(ListNode* head) {
        return f(head, nullptr);
    }
};